### PR TITLE
quic: add listener config option to attach connection debug visitors

### PR DESCRIPTION
**Pull request by wlampert**
Commit Message: Add a way to configure a quic connection debug visitor factory that will be used to attach a debug visitor to all quic connections on the listener. Adds an interface for this new type of factory.
Additional Description:
Risk Level: Low
Testing: Added new tests and modified existing tests in /test/common/quic. Also performed manual testing on a real machine and sent traffic to it using quic_client.
Docs Changes: Update envoy.config.listener.v3.quic_config.proto inline.
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
FYI: It looks like there is a docs issue:

https://dev.azure.com/cncf/envoy/_build/results?buildId=170030&view=logs&jobId=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=f871f0ce-3dc6-5bd2-7916-5ec61dbe723c

File: source/common/quic/envoy_quic_connection_debug_visitor_factory_interface.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#include "envoy/config/typed_config.h"
+#include "envoy/server/process_context.h"
+
+#include "quiche/quic/core/quic_connection.h"
+#include "quiche/quic/core/quic_session.h"
+
+namespace Envoy {
+namespace Quic {
+
+class EnvoyQuicConnectionDebugVisitorFactoryInterface : public Config::TypedFactory {
```
Comment (RyanTheOptimist) [inline source/common/quic/envoy_quic_connection_debug_visitor_factory_interface.h:20]: This name has a hilarious number of components :) I particularly enjoy the "VisitorFactoryInterface" suffix which is three different design patterns in a row. Remind me of my favorite class in Chrome `HttpStreamFactoryImplJobController`. Anyway, this is fine and the name makes sense. I just found it amusing :)

**Review by RyanTheOptimist — COMMENTED** [review]
Looks great! How plausible would it be to add an integration test to `quic_http_integration_test.cc`? That'd be a good way to confirm that the plumbing all works as expected.

File: source/common/quic/envoy_quic_connection_debug_visitor_factory_interface.h (hunk 1 above)
Comment (wlampert) [reply source/common/quic/envoy_quic_connection_debug_visitor_factory_interface.h:18]: Yeah its a bit ridiculous 😄

**Comment by RyanTheOptimist** [comment]
Looks like an MSAN failure:
```
[  FAILED  ] QuicHttpIntegrationTests/QuicHttpIntegrationTest.ConnectionDebugVisitor/IPv4, where GetParam() = 4-byte object <00-00 00-00>
```

**Review by RyanTheOptimist — APPROVED** [review]
Nice! This looks great.

**Comment by RyanTheOptimist** [comment]
@wbpcode friendly ping for the API review.

**Review by wbpcode — COMMENTED** [review]
/lgtm api

**Comment by alyssawilk** [comment]
This looks like it's causing flakes - Will can you triage?

**Pull request merged**